slice/internal/controller/workload_controller.go

```diff
@@ -64,34 +85,74 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	log.V(2).Info("Reconcile Workload")
 
 	if r.shouldFinalize(wl) {
-		if controllerutil.ContainsFinalizer(wl, CleanupSliceFinalizerName) {
+		if controllerutil.ContainsFinalizer(wl, SliceControllerName) {
 			err = r.client.Delete(ctx, r.newEmptySlice(wl))
 			if client.IgnoreNotFound(err) != nil {
 				return ctrl.Result{}, err
 			}
-			controllerutil.RemoveFinalizer(wl, CleanupSliceFinalizerName)
+			controllerutil.RemoveFinalizer(wl, SliceControllerName)
 			if err := r.client.Update(ctx, wl); err != nil {
-				return ctrl.Result{}, err
+				if !apierrors.IsNotFound(err) {
+					log.Error(err, "Failed to remove finalizer")
+				}
+				return ctrl.Result{}, client.IgnoreNotFound(err)
 			}
 			log.V(5).Info("Removed finalizer")
 		}
 		return ctrl.Result{}, nil
 	}
 
-	if controllerutil.AddFinalizer(wl, CleanupSliceFinalizerName) {
-		if err := r.client.Update(ctx, wl); err != nil {
-			log.V(5).Info("Added finalizer")
-			return ctrl.Result{}, err
+	ac, err := r.sliceAC(ctx, wl)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if ac == nil {
+		log.V(5).Info("Admission check not found – ignoring reconciliation for now")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("admissionCheck", ac.Name)
+	ctrl.LoggerInto(ctx, log)
+
+	if ac.State == kueue.CheckStateReady {
+		log.V(5).Info("Admission check is ready — nothing to do")
+		return reconcile.Result{}, nil
+	}
+
+	if controllerutil.AddFinalizer(wl, SliceControllerName) {
+		if err = r.client.Update(ctx, wl); err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "Failed to add finalizer")
+			}
+			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
+		log.V(5).Info("Added finalizer")
+		return ctrl.Result{}, nil
 	}
 
-	return ctrl.Result{}, r.createSliceIfNotExist(ctx, wl)
+	err = r.createSliceIfNotExist(ctx, wl, ac)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, client.IgnoreNotFound(r.handleSuccess(ctx, wl, ac))
 }
 
 func (r *WorkloadReconciler) shouldFinalize(wl *kueue.Workload) bool {
 	return !wl.DeletionTimestamp.IsZero() || workload.IsFinished(wl) || workload.IsEvicted(wl) || !workload.IsActive(wl)
 }
 
+func (r *WorkloadReconciler) sliceAC(ctx context.Context, wl *kueue.Workload) (*kueue.AdmissionCheckState, error) {
+	relevantChecks, err := admissioncheck.FilterForController(ctx, r.client, wl.Status.AdmissionChecks, SliceControllerName)
+	if err != nil {
+		return nil, err
+	}
+	if len(relevantChecks) == 0 {
+		return nil, nil
+	}
```

Comment:
Should there be exactly one? If yes, should we assert that that is the case?

If we have >1 slice for the workload, will we just use 1 AC?

Reply:
Yeah, you right. Than I think we should merge https://github.com/AI-Hypercomputer/xpk/pull/520 first.

Reply:
As we discussed, we don’t need to have multiple admission checks.

Reply:
Not sure, to be honest. Maybe we should create a webhook to validate whether an AdmissionCheck already exists. What do you think?